### PR TITLE
soft_margin_loss bug fix (#80206)

### LINKED ISSUE #80206: `soft_margin_loss` gives wrong gradient when `target` with dtype uint8
### 🐛 Describe the bug

`soft_margin_loss`gives wrong gradient when`target` with dtype uint8

```python
import torch

torch.random.manual_seed(9782)
def get_fn():
    target = torch.tensor([0, 0, 1, 1], dtype=torch.uint8, device='cpu')
    def fn(input):
        fn_res = torch.nn.functional.soft_margin_loss(input, target)
        return fn_res
    return fn
fn = get_fn()
input = torch.tensor([[0.9707, 0.2672, 0.2389, 0.6876]], dtype=torch.float64, device='cpu', requires_grad=True)
try:
    torch.autograd.gradcheck(fn, (input), check_sparse_nnz=False, atol=0.01, rtol=0.01, check_forward_ad=False, check_backward_ad=True, check_batched_grad=False)
except Exception as e:
    print(e)

# Jacobian mismatch for output 0 with respect to input 0,
# numerical:tensor([[ 0.0000],
#         [ 0.0000],
#         [-0.1101],
#         [-0.0836]], dtype=torch.float64)
# analytical:tensor([[-0.0000],
#         [-0.0000],
#         [-0.2500],
#         [-0.2500]], dtype=torch.float64)
```

If `target` has dtype `int16` or `int32`, it will return the correct gradient.

### Versions

pytorch: 1.11.0

### PR DESCRIPTION
**Pull request by cybershiptrooper**
Fixes #80206

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @saketh-are

### CODE REVIEW

**Comment by cybershiptrooper** [comment]
Hi, kinda new here- had a couple of questions regarding this:
1. The python file in the issue is working. Do I need to add/check any other test cases for this?
2. The current [documentation](https://pytorch.org/docs/master/generated/torch.nn.SoftMarginLoss.html#torch.nn.SoftMarginLoss) asks to put only -1 or 1..

**Comment by ezyang** [comment]
Can you add the test case to the test suite. Maybe test/test_nn.py or test_torch.py

I also think disallowing uint8 is a good idea, and I would add that. If we disallow all unsigned types, you might not even need to change the logic here.

**Comment by cybershiptrooper** [comment]
@ezyang adding a test case(for this commit) and disallowing uint seem contradictory.. Should I remove the commit then?

**Comment by ezyang** [comment]
If you disallow uint, the test would just be that uint input raises error.
